### PR TITLE
Google geocoder

### PR DESCRIPTION
**Pull request by ethervoid**
This PR includes:

- Add google geocoder logic
- Refactored the python library to change to a services mode because we are going to start including more services like routing

### CODE REVIEW

**Comment by ethervoid** [comment]
retest this please

**Comment by ethervoid** [comment]
retest this please

**Comment by ethervoid** [comment]
retest this please

**Comment by ethervoid** [comment]
@rafatower check this out

**Comment by ethervoid** [comment]
retest this please

File: server/extension/cdb_geocoder_server--0.1.0--0.2.0.sql [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+      heremaps_app_id = None
+      heremaps_app_code = None
+    else:
+      heremaps_conf = json.loads(heremaps_conf_json)
+      heremaps_app_id = heremaps_conf['app_id']
+      heremaps_app_code = heremaps_conf['app_code']
+    geocoder_config = GeocoderConfig(redis_conn, username, orgname, heremaps_app_id, heremaps_app_code)
+    # --Think about the security concerns with this kind of global cache, it should be only available
+    # --for this user session but...
+    GD[cache_key] = geocoder_config
```
Comment (rafatower) [inline server/extension/cdb_geocoder_server--0.1.0--0.2.0.sql]: As we cannot guarantee the session will not be shared and `geocoder_config` var is dependent on the user/org, please store it in `SD` dict.

http://www.postgresql.org/docs/9.3/static/plpython-sharing.html
https://github.com/plproxy/plproxy/blob/master/doc/syntax.md#connect

File: server/extension/cdb_geocoder_server--0.1.0--0.2.0.sql [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        metadata_config_params['sentinel_port'],
+        metadata_config_params['sentinel_master_id'],
+        timeout=metadata_config_params['timeout'],
+        redis_db=metadata_config_params['redis_db']).redis_connection()
+    redis_metrics_connection = RedisConnection(metrics_config_params['sentinel_host'],
+        metrics_config_params['sentinel_port'],
+        metrics_config_params['sentinel_master_id'],
+        timeout=metrics_config_params['timeout'],
+        redis_db=metrics_config_params['redis_db']).redis_connection()
+    GD[cache_key] = {
```
Comment (rafatower) [inline server/extension/cdb_geocoder_server--0.1.0--0.2.0.sql]: same here

Comment (rafatower) [inline server/extension/cdb_geocoder_server--0.1.0--0.2.0.sql]: no, not here but in anything in `GD` that is user-dependent

File: server/extension/cdb_geocoder_server--0.1.0--0.2.0.sql (hunk 1 above)
Comment (rafatower) [inline server/extension/cdb_geocoder_server--0.1.0--0.2.0.sql]: Ok, all of those as long as global stuff is accessed per-user and not stored globally.

File: server/extension/cdb_geocoder_server--0.1.0--0.2.0.sql [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+      return None
+  except BaseException as e:
+    import sys, traceback
+    type_, value_, traceback_ = sys.exc_info()
+    quota_service.increment_failed_geocoder_use()
+    error_msg = 'There was an error trying to geocode using here maps geocoder: {0}'.format(e)
+    plpy.notice(traceback.format_tb(traceback_))
+    plpy.error(error_msg)
+  finally:
+    quota_service.increment_total_geocoder_use()
```
Comment (rafatower) [inline server/extension/cdb_geocoder_server--0.1.0--0.2.0.sql]: nice!

File: server/extension/sql/0.2.0/40_admin1.sql [hunk 4]
```diff
@@ -0,0 +1,89 @@
+-- Interfacess of the server extension
```
Comment (rafatower) [inline server/extension/sql/0.2.0/40_admin1.sql]: typo

File: server/extension/sql/0.2.0/50_namedplaces.sql [hunk 5]
```diff
@@ -0,0 +1,121 @@
+-- Interfacess of the server extension
```
Comment (rafatower) [inline server/extension/sql/0.2.0/50_namedplaces.sql]: typo

File: server/lib/python/cartodb_services/cartodb_services/google/geocoder.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        if city:
+            optional_params['locality'] = city
+        if state:
+            optional_params['administrative_area'] = state
+        if country:
+            optional_params['country'] = country
+        return optional_params
+
+    def _clean_client_id(self, client_id):
+        # Consistency with how the client_id is saved in metadata
```
Comment (rafatower) [inline server/lib/python/cartodb_services/cartodb_services/google/geocoder.py]: :-/ but ok

File: server/lib/python/cartodb_services/requirements.txt [hunk 7]
```diff
@@ -1,8 +1,13 @@
 redis==2.10.5
+hiredis==0.1.5
 # Dependency with incsv in the import
 python-dateutil==2.2
+googlemaps==2.4.2
+# Dependency for googlemaps package
+requests<=2.9.1
```
Comment (rafatower) [inline server/lib/python/cartodb_services/requirements.txt]: `<=` ??

Comment (rafatower) [inline server/lib/python/cartodb_services/requirements.txt]: ok to be added to chef

maybe set 2 <= requests <= 2.9 to be on the safe (don't know about the changes in the api)

File: server/lib/python/cartodb_services/test/test_heremapsgeocoder.py [hunk 8]
```diff
@@ -65,7 +68,7 @@ class GeocoderTestCase(unittest.TestCase):
                             "District": "Trafalgar",
                             "Street": "Calle de Eloy Gonzalo",
                             "AdditionalData": [{
-                                "value": "España",
+                                "value": "Espana",
```
Comment (rafatower) [inline server/lib/python/cartodb_services/test/test_heremapsgeocoder.py]: :(
(just curious: any reason for this? I guess you could use `u"""España"""` but not worth changing it)

**Comment by rafatower** [comment]
In general ok, my comments are just small details

Sorry about commenting on generated files. Small suggestion: add 
```sql
--DO NOT MODIFY THIS FILE, IT IS GENERATED AUTOMATICALLY FROM SOURCES
```
to the header template file.

File: server/lib/python/cartodb_services/requirements.txt (hunk 7 above)
Comment (ethervoid) [inline server/lib/python/cartodb_services/requirements.txt]: Done, is controlled by chef. I want to have it this way like googlemaps has it.

**Comment by ethervoid** [comment]
You are talking about the cdb_geocoder_server--x.y.z.sql files aren't you?

**Comment by rafatower** [comment]
I was talking about adding that line to these files:
https://github.com/CartoDB/geocoder-api/blob/master/server/extension/sql/0.1.0/00_header.sql
https://github.com/CartoDB/geocoder-api/blob/master/client/sql/0.1.0/00_header.sql

**Comment by ethervoid** [comment]
Are you sure? That files are not automatically generated...

**Comment by ethervoid** [comment]
retest this please

**Pull request merged**